overlord/configstate/configcore/services.go

```diff
@@ -231,6 +239,147 @@ func handleServiceDisableConfiguration(_ sysconfig.Device, tr config.ConfGetter,
+	// error when no host is given, so workaround here
+	if host == "" && portStr != "" {
+		return []string{
+			fmt.Sprintf("0.0.0.0:%v", portStr),
+			fmt.Sprintf("[::]:%v", portStr),
+		}, nil
+	}
+
+	// no special handling needed and a valid listen address
+	return []string{oneAddr}, nil
+}
+
+func parseSSHListenCfg(cfgStr string) ([]string, error) {
+	var listenAddrs []string
+	for _, hostAndPort := range strings.Split(cfgStr, ",") {
+		addrs, err := parseOneSSHListenAddr(hostAndPort)
+		if err != nil {
+			return nil, err
+		}
+		listenAddrs = append(listenAddrs, addrs...)
+	}
+	return listenAddrs, nil
+}
+
+func validateServiceConfiguration(tr config.ConfGetter) error {
+	// validate the ssh listen setting
+	output, err := coreCfg(tr, sshListenOpt)
+	if err != nil {
+		return err
+	}
+	if output == "" {
+		return nil
+	}
+	if _, err := parseSSHListenCfg(output); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func handleServiceConfigSSHListen(dev sysconfig.Device, tr config.ConfGetter, opts *fsOnlyContext) error {
+	// see if anything needs to happenhan
+	var pristineSSHListen, newSSHListen interface{}
+
+	if err := tr.GetPristine("core", sshListenOpt, &pristineSSHListen); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if err := tr.Get("core", sshListenOpt, &newSSHListen); err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if pristineSSHListen == newSSHListen {
+		return nil
+	}
+
+	// ssh.port config has changed, write new config
+	root := dirs.GlobalRootDir
+	if opts != nil {
+		root = opts.RootDir
+	}
+
+	// Note: Only UC20+ supports using the "sshd_config.d"
+	// dir. Supporting older systems would be hard because we
+	// would have to merge somehow the UC16 and UC18 config in a
+	// fsOnlyContext
+	if !dev.HasModeenv() {
+		return fmt.Errorf("cannot set ssh listen address configuration on systems older than UC20")
+	}
+
+	name := "listen.conf"
+	dirContent := map[string]osutil.FileState{}
+	if newSSHListen != nil && newSSHListen != "" {
+		listenAddrs, err := parseSSHListenCfg(fmt.Sprintf("%v", newSSHListen))
+		if err != nil {
+			return fmt.Errorf("cannot set ssh configuration: %v", err)
+		}
+
+		var buf bytes.Buffer
+		for _, s := range listenAddrs {
+			if _, err := fmt.Fprintf(&buf, "ListenAddress %v\n", s); err != nil {
+				return fmt.Errorf("cannot create ssh option buffer: %v", err)
+			}
+		}
+
+		dirContent[name] = &osutil.MemoryFileState{
+			Content: buf.Bytes(),
+			Mode:    0600,
+		}
+	}
+	dir := filepath.Join(root, "/etc/ssh/sshd_config.d/")
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return err
+	}
+	_, _, err := osutil.EnsureDirState(dir, name, dirContent)
+	if err != nil {
+		return err
+	}
+
+	var sysd systemd.Systemd
+	if opts != nil {
+		sysd = systemd.NewEmulationMode(opts.RootDir)
+	} else {
+		sysd = systemd.New(systemd.SystemMode, &sysdLogger{})
+	}
+
+	if err := sysd.ReloadOrRestart("ssh.service"); err != nil {
+		return err
```

Comment:
wasn't it simpler to do that what swap.go does and call Restart only in non emulated case, or am I missing something?